internal/services/azure_reservation_service.go

```diff
@@ -0,0 +1,133 @@
+	"github.com/RHEnVision/provisioning-backend/internal/payloads"
+	"github.com/RHEnVision/provisioning-backend/internal/queue"
+	"github.com/RHEnVision/provisioning-backend/pkg/worker"
+	"github.com/go-chi/render"
+)
+
+func CreateAzureReservation(w http.ResponseWriter, r *http.Request) {
+	logger := *ctxval.Logger(r.Context())
+
+	payload := &payloads.AzureReservationRequestPayload{}
+	if err := render.Bind(r, payload); err != nil {
+		renderError(w, r, payloads.NewInvalidRequestError(r.Context(), "Azure reservation", err))
+		return
+	}
+
+	pkDao := dao.GetPubkeyDao(r.Context())
+	rDao := dao.GetReservationDao(r.Context())
+
+	// validate pubkey
+	logger.Debug().Msgf("Validating existence of pubkey %d for this account", payload.PubkeyID)
+	pk, err := pkDao.GetById(r.Context(), payload.PubkeyID)
+	if err != nil {
+		message := fmt.Sprintf("get pubkey with id %d", payload.PubkeyID)
+		renderNotFoundOrDAOError(w, r, err, message)
+		return
+	}
+	logger.Debug().Msgf("Found pubkey %d named '%s'", pk.ID, pk.Name)
+
+	// Get Sources client
+	sourcesClient, err := clients.GetSourcesClient(r.Context())
+	if err != nil {
+		renderError(w, r, payloads.NewClientError(r.Context(), err))
+		return
+	}
+
+	// Fetch SubscriptionID from Sources
+	authentication, err := sourcesClient.GetAuthentication(r.Context(), payload.SourceID)
+	if err != nil {
+		renderError(w, r, payloads.NewClientError(r.Context(), err))
+		return
+	}
+
+	if typeErr := authentication.MustBe(models.ProviderTypeAzure); typeErr != nil {
+		renderError(w, r, payloads.NewClientError(r.Context(), typeErr))
+		return
+	}
+
+	var azureImageID string
+	if strings.HasPrefix(payload.ImageID, "/subscriptions/") {
+		// Direct AMI ID was provided, no need to call image builder
+		azureImageID = payload.ImageID
+	} else {
+		// TODO fetch image ID from Image Builder
+		renderError(w, r, payloads.NewInvalidRequestError(r.Context(), "fetching Azure image id from Image Builder is not implemented", ProviderTypeNotImplementedError))
+		return
+	}
+
+	supportedArch := "x86_64"
+	it := types.FindInstanceSize(clients.InstanceTypeName(payload.InstanceSize))
+	if it == nil {
+		renderError(w, r, payloads.NewInvalidRequestError(r.Context(), fmt.Sprintf("unknown instance size: %s", payload.InstanceSize), UnknownInstanceTypeNameError))
+		return
+	}
+	if it.Architecture.String() != supportedArch {
+		renderError(w, r, payloads.NewWrongArchitectureUserError(r.Context(), ArchitectureMismatch))
+		return
+	}
+
+	name := config.Application.InstancePrefix + payload.Name
+	detail := &models.AzureDetail{
+		Location:     payload.Location,
+		InstanceSize: payload.InstanceSize,
+		Amount:       payload.Amount,
+		PowerOff:     payload.PowerOff,
+		Name:         name,
+	}
+	reservation := &models.AzureReservation{
+		PubkeyID: payload.PubkeyID,
+		SourceID: payload.SourceID,
+		ImageID:  payload.ImageID,
+		Detail:   detail,
+	}
+	reservation.Steps = 1
+	reservation.StepTitles = []string{"Launch instance(s)"}
```

Comment:
This hurts, please add all of the 51 steps that are required in order to launch Azure VM so users get some idea on what is going on when something is not responding.

Reply:
mhm... could we do it in a follow up? That will require a LOT of work as the separate functions are not exported...

Reply:
Will I get away with https://issues.redhat.com/browse/HMS-1165 ? :)

Reply:
Sure, I was thinking in this PR because it is a new code so no problem, but we already had a conflict so...